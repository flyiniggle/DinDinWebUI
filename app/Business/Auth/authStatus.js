const TOKEN_KEY = 'DinDinLoggedIn';
const USERNAME_KEY = 'DinDinUser';

const authStatus = {
    get authToken() {
        return window.localStorage.getItem(TOKEN_KEY);
    },

    set authToken(token) {
        if (token === '') {
            window.localStorage.removeItem(TOKEN_KEY);
        }
        window.localStorage.setItem(TOKEN_KEY, token.toString());
    },

    get username() {
        return window.localStorage.getItem(USERNAME_KEY) || 'not logged in';
    },

    set username(username) {
        if (username === '') {
            window.localStorage.removeItem(USERNAME_KEY);
        }
        return window.localStorage.setItem(USERNAME_KEY, username);
    }
};

authStatus.logOut = function() {
    this.authToken = '';
    this.username = '';
}.bind(authStatus);

export default Object.seal(authStatus);
export { TOKEN_KEY, USERNAME_KEY };