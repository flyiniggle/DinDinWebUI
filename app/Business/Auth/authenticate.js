import authStatus from 'Business/Auth/authStatus';
import AuthService from 'Business/Auth/Service';
import preflightCheck from 'Business/Auth/Validation/preflightCheck';
import responseCheck from 'Business/Auth/Validation/responseCheck';
import { Result } from 'true-myth';

// String => String => Promise(Response.json())
async function authenticate(username, password) {
    const errors = preflightCheck({username, password});

    if (errors.length > 0) {
        return Result.err(errors);
    }

    const result = await AuthService.get(username, password);

    if (result.isErr()) {
        result.mapErr(responseCheck);
    } else {
        authStatus.loggedIn = true;
        return result;
    }
}

export default authenticate;