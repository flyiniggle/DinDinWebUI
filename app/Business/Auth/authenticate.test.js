import authenticate from 'Business/Auth/authenticate';
import authStatus from 'Business/Auth/authStatus';
import { fake, replace, restore } from 'sinon';
import AuthService from 'Business/Auth/Service';


describe('#Business #Auth #suthenticate', function() {
    it('should reject if no username is provided.', function() {
        expect(authenticate()).rejects.toMatchSnapshot();
    });

    it('should reject if no password is provided.', function() {
        expect(authenticate(test)).rejects.toMatchSnapshot();
    });

    it('should reject if it recieves a failed credentials validation message from the server', function() {
        const serviceFake = fake.resolves({non_field_errors: ['Unable to log in with provided credentials.']});

        replace(AuthService, 'get', serviceFake);

        expect(authenticate('user', 'password')).rejects.toMatchSnapshot();
        restore();
    });

    it('should reject if it receives an unknown validation message from the server', function() {
        const serviceFake = fake.resolves({someError: ['this is an error.']});

        replace(AuthService, 'get', serviceFake);

        expect(authenticate('user', 'password')).rejects.toMatchSnapshot();
        restore();
    });

    it('should reject if it receives no token', function() {
        const serviceFake = fake.resolves({});

        replace(AuthService, 'get', serviceFake);

        expect(authenticate('user', 'password')).rejects.toMatchSnapshot();
        restore();
    });

    it('should resolve to the token if successful', function() {
        const serviceFake = fake.resolves({token: 123456});

        replace(AuthService, 'get', serviceFake);

        expect(authenticate('user', 'password')).resolves.toEqual({token: 123456});
        restore();
    });

    it('should set authStatus.loggedIn to true if successful.', async function() {
        const serviceFake = fake.resolves({token: 123456});

        replace(AuthService, 'get', serviceFake);

        expect.assertions(1)

        await authenticate('user', 'password');
        expect(authStatus.loggedIn).toEqual(true);
        restore();
    });
});