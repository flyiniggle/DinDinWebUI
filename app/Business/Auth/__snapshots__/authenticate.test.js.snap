// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Business #Auth #suthenticate should reject if it receives an unknown validation message from the server 1`] = `
Array [
  Message {
    "field": "password",
    "message": "It is not possible to log in right now.",
    "type": "Error",
    "value": "",
  },
]
`;

exports[`#Business #Auth #suthenticate should reject if it receives no token 1`] = `
Array [
  Message {
    "field": "password",
    "message": "It is not possible to log in right now.",
    "type": "Error",
    "value": "",
  },
]
`;

exports[`#Business #Auth #suthenticate should reject if it recieves a failed credentials validation message from the server 1`] = `
Array [
  Message {
    "field": "password",
    "message": "Username and password did not match.",
    "type": "Error",
    "value": "",
  },
]
`;

exports[`#Business #Auth #suthenticate should reject if no password is provided. 1`] = `
Array [
  Message {
    "field": "password",
    "message": "required",
    "type": "Error",
    "value": undefined,
  },
]
`;

exports[`#Business #Auth #suthenticate should reject if no username is provided. 1`] = `
Array [
  Message {
    "field": "username",
    "message": "required",
    "type": "Error",
    "value": undefined,
  },
  Message {
    "field": "password",
    "message": "required",
    "type": "Error",
    "value": undefined,
  },
]
`;
